File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/iterator/concrete_iterator.dart
import 'iterator.dart';
import 'concrete_aggregate.dart';
import 'types.dart';

class PlaylistIterator implements IteratorIface<Song> {
  int _index = 0;
  final Playlist aggregate;

  PlaylistIterator(this.aggregate);

  @override
  bool hasNext() => _index < aggregate.length;

  @override
  Song next() {
    if (!hasNext()) throw Exception('No more elements');
    return aggregate.getAt(_index++);
  }

  @override
  void remove() {
    if (_index == 0) throw Exception('Call next() before remove()');
    aggregate.removeAt(_index - 1);
    _index--;
  }

  @override
  void reset() {
    _index = 0;
  }
}

File: implementations/react/src/patterns/behavioral/iterator/concrete-iterator.ts
import type { Playlist } from "./concrete-aggregate";
import type { Song } from "./types/type";
import type { Iterator } from "./iterator";

/**
 * Iterador concreto para la lista de reproducción de canciones.
 */

export class PlaylistIterator implements Iterator<Song> {
  private index = 0;
  private readonly aggregate: Playlist;
  constructor(aggregate: Playlist) {
    this.aggregate = aggregate;
  }

  hasNext(): boolean {
    return this.index < this.aggregate.length;
  }
  next(): Song {
    if (!this.hasNext()) throw new Error("No more elements");
    return this.aggregate.getAt(this.index++);
  }
  remove(): void {
    if (this.index === 0) throw new Error("Call next() before remove()");
    // remove the item returned by the last next()
    this.aggregate.removeAt(this.index - 1);
    this.index--; // step back so iteration stays in sync
  }
  reset(): void { this.index = 0; }
}